Common/Data/DividendYieldProvider.cs

```diff
@@ -0,0 +1,192 @@
+            {
+                var dueTime = Time.GetNextLiveAuxiliaryDataDueTime();
+                if (dueTime > TimeSpan.FromMinutes(10))
+                {
+                    // Clear the cache before the auxiliary due time to avoid race conditions with consumers
+                    return dueTime - TimeSpan.FromMinutes(10);
+                }
+                return dueTime;
+            }
+        }
+
+        /// <summary>
+        /// Instantiates a <see cref="DividendYieldProvider"/> with the specified Symbol
+        /// </summary>
+        public DividendYieldProvider(Symbol symbol)
+        {
+            _symbol = symbol;
+            StartExpirationTask();
+        }
+
+        /// <summary>
+        /// Helper method that will clear any cached dividend rate in a daily basis, this is useful for live trading
+        /// </summary>
+        protected virtual void StartExpirationTask()
+        {
+            lock (_lock)
+            {
+                // we clear the dividend yield rate cache so they are reloaded
+                _dividendYieldRateProvider = null;
+            }
+            _ = Task.Delay(CacheRefreshPeriod).ContinueWith(_ => StartExpirationTask());
+        }
+
+        /// <summary>
+        /// Lazily loads the dividend provider from disk and returns it
+        /// </summary>
+        private IReadOnlyDictionary<DateTime, decimal> DividendYieldRateProvider
+        {
+            get
+            {
+                // let's not lock if the provider is already loaded
+                if (_dividendYieldRateProvider != null)
+                {
+                    return _dividendYieldRateProvider;
+                }
+
+                lock (_lock)
+                {
+                    if (_dividendYieldRateProvider == null)
+                    {
+                        LoadDividendYieldProvider();
+                    }
+                    return _dividendYieldRateProvider;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get dividend yield by a given date
+        /// </summary>
+        /// <param name="date">The date</param>
+        /// <returns>Dividend yield on the given date</returns>
+        public decimal GetDividendYield(DateTime date)
+        {
+            if (!DividendYieldRateProvider.TryGetValue(date.Date, out var dividendYield))
+            {
+                return date < _firstDividendYieldDate
+                    ? DefaultDividendYieldRate
+                    : _lastDividendYield;
+            }
+
+            return dividendYield;
+        }
+
+        /// <summary>
+        /// Generate the daily historical dividend yield
+        /// </summary>
+        /// <remarks>Exposed for testing</remarks>
+        protected virtual void LoadDividendYieldProvider()
+        {
+            var factorFileProvider = Composer.Instance.GetPart<IFactorFileProvider>();
+            var corporateFactors = factorFileProvider
+                .Get(_symbol)
```

Comment:
Whops, here ur loading the factors only for the set symbol, what about the others, given the `_dividendYieldRateProvider` is static? Seems it should be a per symbol dictionary being used? like the map/factor file providers use